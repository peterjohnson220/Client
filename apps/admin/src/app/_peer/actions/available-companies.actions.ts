import { Action } from '@ngrx/store';

import { AvailableCompany } from 'libs/models/peer';
import { GridDataResult } from '@progress/kendo-angular-grid';

export const LOADING_AVAILABLE_COMPANIES  = '[Peer Admin/Available Companies] Loading Available Companies';
export const LOADING_AVAILABLE_COMPANIES_SUCCESS  = '[Peer Admin/Available Companies] Loading Available Companies Success';
export const LOADING_AVAILABLE_COMPANIES_ERROR  = '[Peer Admin/AVAILABLE Companies] Loading Available Companies Error';

export class LoadingAvailableCompanies implements Action {
  readonly type = LOADING_AVAILABLE_COMPANIES;

  // TODO: model?
  constructor(public payload: any) {}
}

export class LoadingAvailableCompaniesSuccess implements Action {
  readonly type = LOADING_AVAILABLE_COMPANIES_SUCCESS;

  // TODO: Need strong type?
  constructor(public payload: GridDataResult) {}
}

export class LoadingAvailableCompaniesError implements Action {
  readonly type = LOADING_AVAILABLE_COMPANIES_ERROR;
}

export type Actions
  = LoadingAvailableCompanies
  | LoadingAvailableCompaniesSuccess
  | LoadingAvailableCompaniesError;




