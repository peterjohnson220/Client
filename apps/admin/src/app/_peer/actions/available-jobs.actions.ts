import { Action } from '@ngrx/store';
import { GridDataResult } from '@progress/kendo-angular-grid';

export const LOADING_AVAILABLE_JOBS  = '[Peer Admin/Available Jobs] Loading Available Jobs';
export const LOADING_AVAILABLE_JOBS_SUCCESS  = '[Peer Admin/Available Jobs] Loading Available Jobs Success';
export const LOADING_AVAILABLE_JOBS_ERROR  = '[Peer Admin/Available Jobs] Loading Available Jobs Error';

export class LoadingAvailableJobs implements Action {
  readonly type = LOADING_AVAILABLE_JOBS;

  constructor(public payload: any) {}
}

export class LoadingAvailableJobsSuccess implements Action {
  readonly type = LOADING_AVAILABLE_JOBS_SUCCESS;

  constructor(public payload: GridDataResult) {}
}

export class LoadingAvailableJobsError implements Action {
  readonly type = LOADING_AVAILABLE_JOBS_ERROR;
  readonly payload;
}

export type Actions
  = LoadingAvailableJobs
  | LoadingAvailableJobsSuccess
  | LoadingAvailableJobsError;
