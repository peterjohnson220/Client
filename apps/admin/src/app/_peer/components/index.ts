export * from './create-exchange-modal.component';
