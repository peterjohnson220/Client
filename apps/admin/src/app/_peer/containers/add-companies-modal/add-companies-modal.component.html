<pf-modal-form
  [size]="'lg'"
  [title]="'Add Companies'"
  [primaryButtonText]="primaryButtonText"
  [primaryButtonTextSubmitting]="primaryButtonTextSubmitting"
  [submitting]="addingCompanies$ | async"
  [isOpen$]="addCompaniesModalOpen$"
  [formGroup]="addCompaniesForm"
  (onSubmit)="handleFormSubmit()"
  (onDismiss)="handleModalDismissed()">

  <ng-container form-content>

    <pf-async-container
      [loading]="availableCompaniesLoading$ | async"
      [loadingError]="availableCompaniesLoadingError$ | async"
      [loadingErrorMessage]="'Error loading exchange companies'"
      (reload)="loadAvailableCompanies()">

      <div class="form-group">
        <pf-input-debounce id="search-box"
                           [focus]="true"
                           [delay]="400"
                           [(ngModel)]="searchTerm"
                           [ngModelOptions]="{standalone: true}"
                           [placeholderText]="'Search for a Company Name...'"
                           (valueChanged)="updateSearchFilter($event)">
        </pf-input-debounce>
      </div>

      <kendo-grid
        class="row-click-selects no-wrap-ellipsis-values"
        [data]="view$ | async"
        [pageSize]="(gridState$ | async)?.take"
        [skip]="(gridState$ | async)?.skip"
        [resizable]="true"
        [pageable]="{
          buttonCount: 5,
          info: true
        }"
        [selectable]="{
          checkboxOnly: true
        }"
        [height]="400"
        [sortable]="true"
        [sort]="(gridState$ | async)?.sort"
        [kendoGridSelectBy]="selectionKey"
        [selectedKeys]="selections$ | async"
        [rowClass]="rowClass"
        (pageChange)="handlePageChange($event)"
        (sortChange)="handleSortChange($event)"
        (cellClick)="handleCellClick($event)"
      >

        <kendo-grid-column field="CompanyName" title="Company Name" [width]="200"></kendo-grid-column>

        <kendo-grid-column field="Industry" title="Industry" [width]="100"></kendo-grid-column>

        <kendo-grid-column field="Revenue" title="Revenue">
          <ng-template kendoGridCellTemplate let-dataItem>
            {{dataItem.Revenue | humanizeNumber}}
          </ng-template>
        </kendo-grid-column>

        <kendo-grid-column field="FTEs" title="FTEs"></kendo-grid-column>

        <kendo-grid-column field="City" title="City"></kendo-grid-column>

        <kendo-grid-column field="State" title="State"></kendo-grid-column>

        <kendo-grid-column field="Zip" title="Zip"></kendo-grid-column>

      </kendo-grid>

    </pf-async-container>

    <div class="form-group hidden-form-control">
      <input type="hidden" class="form-control" name="hasSelections" formControlName="selections" [ngModel]="selections$ | async" [pfValidatable]="attemptedSubmit"/>
      <ng-container *ngIf="selectionsControl.errors && attemptedSubmit">
        <div *ngIf="selectionsControl.errors.error" class="invalid-feedback">{{selectionsControl.errors.error}}</div>
      </ng-container>
    </div>

  </ng-container>

</pf-modal-form>

