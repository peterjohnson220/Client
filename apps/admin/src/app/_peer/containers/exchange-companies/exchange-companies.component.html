<pf-async-container
  [loading]="exchangeCompaniesLoading$ | async"
  [loadingError]="exchangeCompaniesLoadingError$ | async"
  [loadingErrorMessage]="'Error loading exchange companies'"
  (reload)="handleExchangeCompaniesGridReload()">

  <kendo-grid [data]="exchangeCompanies$ | async">

    <kendo-grid-column field="CompanyName" title="Company Name"></kendo-grid-column>

    <kendo-grid-column field="Industry" title="Industry"></kendo-grid-column>

    <kendo-grid-column field="JobCount" title="Job Count"></kendo-grid-column>

    <kendo-grid-column field="Location" title="Location"></kendo-grid-column>

    <ng-template kendoPagerTemplate let-totalPages="totalPages" let-currentPage="currentPage">
      <kendo-pager-info></kendo-pager-info>
    </ng-template>

  </kendo-grid>

</pf-async-container>
