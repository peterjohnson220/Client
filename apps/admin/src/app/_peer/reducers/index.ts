import { createSelector, createFeatureSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromManageExchangeReducer from './manage-exchange.reducer';
import * as fromExchangeCompaniesReducer from './exchange-companies.reducer';
import * as fromImportExchangeJobsReducer from './import-exchange-jobs.reducer';
import * as fromAvailableCompaniesReducer from './available-companies.reducer';

// Feature area state
export interface PeerAdminState {
  manageExchange: fromManageExchangeReducer.State;
  exchangeCompanies: fromExchangeCompaniesReducer.State;
  importExchangeJobs: fromImportExchangeJobsReducer.State;
  availableCompanies: fromAvailableCompaniesReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  peerAdmin: PeerAdminState;
}

// Feature area reducers
export const reducers = {
  manageExchange: fromManageExchangeReducer.reducer,
  exchangeCompanies: fromExchangeCompaniesReducer.reducer,
  importExchangeJobs: fromImportExchangeJobsReducer.reducer,
  availableCompanies: fromAvailableCompaniesReducer.reducer,
};

// Select Feature Area
export const selectPeerAdminState = createFeatureSelector<PeerAdminState>('peerAdmin');

// Feature Selectors
export const selectManageExchangeState = createSelector(selectPeerAdminState, (state: PeerAdminState) => state.manageExchange);
export const selectExchangeCompaniesState = createSelector(selectPeerAdminState, (state: PeerAdminState) => state.exchangeCompanies);
export const selectImportExchangeJobsState = createSelector(selectPeerAdminState, (state: PeerAdminState) => state.importExchangeJobs);
// Manage Exchange Selectors
export const getManageExchange = createSelector(selectManageExchangeState, fromManageExchangeReducer.getExchange);
export const selectAvailableCompaniesState = createSelector(selectPeerAdminState, (state: PeerAdminState) => state.availableCompanies);

// Exchange List Selectors
export const getManageExchangeLoading = createSelector(
  selectManageExchangeState, fromManageExchangeReducer.getLoading
);
export const getManageExchangeLoadingError = createSelector(
  selectManageExchangeState, fromManageExchangeReducer.getLoadingError
);

// Import Exchange Job Selectors
export const getImportExchangeJobsModalOpen = createSelector(
  selectImportExchangeJobsState, fromImportExchangeJobsReducer.getImportExchangeJobsModalOpen
);
export const getImportExchangeJobsUploadingFile = createSelector(
  selectImportExchangeJobsState, fromImportExchangeJobsReducer.getUploadingFile
);
export const getImportExchangeJobsUploadingFileError = createSelector(
  selectImportExchangeJobsState, fromImportExchangeJobsReducer.getUploadingFileError
);
export const getStoredDataFile = createSelector(
  selectImportExchangeJobsState, fromImportExchangeJobsReducer.getStoredDataFile
);
export const getValidationResults = createSelector(
  selectImportExchangeJobsState, fromImportExchangeJobsReducer.getValidationResults
);
export const getImportingJobs = createSelector(
  selectImportExchangeJobsState, fromImportExchangeJobsReducer.getImportingJobs
);
export const getIsFileValid = createSelector(
  selectImportExchangeJobsState, fromImportExchangeJobsReducer.getIsFileValid
);

// Exchange Companies Selectors
export const {
  selectAll: getExchangeCompanies
} = fromExchangeCompaniesReducer.adapter.getSelectors(selectExchangeCompaniesState);

export const getExchangeCompaniesLoading = createSelector(
  selectExchangeCompaniesState, fromExchangeCompaniesReducer.getLoading
);
export const getExchangeCompaniesLoadingError = createSelector(
  selectExchangeCompaniesState, fromExchangeCompaniesReducer.getLoadingError
);

export const getAddExchangeCompaniesModalOpen = createSelector(
  selectExchangeCompaniesState, fromExchangeCompaniesReducer.getAddModalOpen
);

export const getExchangeCompaniesAdding = createSelector(
  selectExchangeCompaniesState, fromExchangeCompaniesReducer.getAdding
);

export const getExchangeCompaniesAddingError = createSelector(
  selectExchangeCompaniesState, fromExchangeCompaniesReducer.getAddingError
);


// Available Companies Selectors
export const {
  selectAll: getAvailableCompanies
} = fromAvailableCompaniesReducer.adapter.getSelectors(selectAvailableCompaniesState);

export const getAvailableCompaniesLoading = createSelector(selectAvailableCompaniesState, fromAvailableCompaniesReducer.getLoading);
export const getAvailableCompaniesLoadingError = createSelector(
  selectAvailableCompaniesState,
  fromAvailableCompaniesReducer.getLoadingError
);
export const getTotalAvailableCompanies = createSelector(selectAvailableCompaniesState, fromAvailableCompaniesReducer.getTotal);
export const getAvailableCompaniesGrid = createSelector(
  getAvailableCompanies,
  getTotalAvailableCompanies,
  (data, total) => {
    return {data: data, total: total};
  }
);
