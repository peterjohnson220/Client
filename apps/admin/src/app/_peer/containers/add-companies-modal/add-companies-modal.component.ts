import { Component, OnDestroy, OnInit, ViewChild } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { FormBuilder, FormGroup } from '@angular/forms';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { Subscription } from 'rxjs/Subscription';
import { GridDataResult, PageChangeEvent, RowArgs, RowClassArgs } from '@progress/kendo-angular-grid';
import { SortDescriptor, State } from '@progress/kendo-data-query';

import { InputDebounceComponent } from 'libs/forms/components';
import { AddExchangeCompaniesRequest } from 'libs/models/peer/index';
import { PfValidators } from 'libs/forms/validators';
import { GridTypeEnum } from 'libs/models/common';
import * as fromGridActions from 'libs/core/actions/grid.actions';

import * as fromAvailableCompaniesActions from '../../actions/available-companies.actions';
import * as fromExchangeCompaniesActions from '../../actions/exchange-companies.actions';
import * as fromPeerAdminReducer from '../../reducers';

@Component({
  selector: 'pf-add-companies-modal',
  templateUrl: './add-companies-modal.component.html',
  styleUrls: ['./add-companies-modal.component.scss']
})
export class AddCompaniesModalComponent implements OnInit, OnDestroy {
  availableCompaniesLoading$: Observable<boolean>;
  availableCompaniesLoadingError$: Observable<boolean>;
  addCompaniesModalOpen$: Observable<boolean>;
  addingCompanies$: Observable<boolean>;
  addingCompaniesError$: Observable<boolean>;
  view$: Observable<GridDataResult>;
  addCompaniesModalOpenSubscription: Subscription;
  addCompaniesErrorSubscription: Subscription;
  addCompaniesForm: FormGroup;
  gridState$: Observable<State>;
  attemptedSubmit = false;
  selections: number[] = [];
  selections$: Observable<number[]>;
  selectionsSubscription: Subscription;
  exchangeId: number;
  searchTerm: string;

  constructor(
    private store: Store<fromPeerAdminReducer.State>,
    private route: ActivatedRoute,
    private fb: FormBuilder
  ) {
    this.availableCompaniesLoading$ = this.store.select(fromPeerAdminReducer.getAvailableCompaniesLoading);
    this.availableCompaniesLoadingError$ = this.store.select(fromPeerAdminReducer.getAvailableCompaniesLoadingError);
    this.view$ = this.store.select(fromPeerAdminReducer.getAvailableCompaniesGrid);
    this.addCompaniesModalOpen$ = this.store.select(fromPeerAdminReducer.getAddExchangeCompaniesModalOpen);
    this.addingCompanies$ = this.store.select(fromPeerAdminReducer.getExchangeCompaniesAdding);
    this.addingCompaniesError$ = this.store.select(fromPeerAdminReducer.getExchangeCompaniesAddingError);
    this.gridState$ = this.store.select(fromPeerAdminReducer.getAvailableCompaniesGridState);
    this.selections$ = this.store.select(fromPeerAdminReducer.getAvailableCompaniesGridSelections);
    this.exchangeId = this.route.snapshot.params.id;
    this.createForm();
  }

  get primaryButtonText() {
    const numOfSelections = this.selections ? this.selections.length : 0;
    return `Add (${numOfSelections})`;
  }
  get primaryButtonTextSubmitting() {
    const numOfSelections = this.selections ? this.selections.length : 0;
    return `Adding (${numOfSelections})`;
  }
  get selectionsControl() { return this.addCompaniesForm.get('selections'); }

  createForm(): void {
    this.addCompaniesForm = this.fb.group({
      'selections': [[], [PfValidators.selectionRequired]]
    });
  }

  // Modal events
  handleFormSubmit(): void {
    this.attemptedSubmit = true;
    const addCompaniesRequest: AddExchangeCompaniesRequest = {
      ExchangeId: this.exchangeId,
      CompanyIds: this.selections
    };
    this.store.dispatch(new fromExchangeCompaniesActions.AddingExchangeCompanies(addCompaniesRequest));
  }

  handleModalDismissed(): void {
    this.attemptedSubmit = false;
    this.searchTerm = '';
    this.store.dispatch(new fromExchangeCompaniesActions.CloseAddExchangeCompaniesModal);
    this.store.dispatch(new fromGridActions.ResetGrid(GridTypeEnum.AvailableCompanies));
  }

  // Grid
  selectionKey(context: RowArgs): number {
    return context.dataItem.CompanyId;
  }

  updateSearchFilter(newSearchTerm: string) {
    this.store.dispatch(new fromGridActions.UpdateFilter(
      GridTypeEnum.AvailableCompanies,
      {columnName: 'CompanyName', value: newSearchTerm}
    ));
    this.loadAvailableCompanies();
  }

  handlePageChange(event: PageChangeEvent): void {
    this.store.dispatch(new fromGridActions.PageChange(GridTypeEnum.AvailableCompanies, event));
    this.loadAvailableCompanies();
  }

  handleSortChange(sort: SortDescriptor[]) {
    this.store.dispatch(new fromGridActions.SortChange(GridTypeEnum.AvailableCompanies, sort));
    this.loadAvailableCompanies();
  }

  handleCellClick(event: any): void {
    if (event.dataItem.InExchange) {
      return;
    }
    const selectedCompanyId = event.dataItem.CompanyId;
    this.store.dispatch(new fromGridActions.ToggleRowSelection(GridTypeEnum.AvailableCompanies, selectedCompanyId));
    this.selectionsControl.markAsTouched();
  }

  rowClass(context: RowClassArgs): string {
     return context.dataItem.InExchange ? 'row-disabled' : '';
  }

  // Lifecycle
  ngOnInit() {
    this.addCompaniesModalOpenSubscription = this.addCompaniesModalOpen$.subscribe(isOpen => {
      if (isOpen) {
        this.loadAvailableCompanies();
      }
    });
    this.addCompaniesErrorSubscription = this.addingCompaniesError$.subscribe(error => {
      if (error) {
        this.selectionsControl.setErrors({'error': 'There was an error adding the selected companies.'});
      }
    });
    this.selectionsSubscription = this.selections$.subscribe(selections => {
      this.selections = selections;
    });
  }

  ngOnDestroy() {
    this.addCompaniesModalOpenSubscription.unsubscribe();
    this.addCompaniesErrorSubscription.unsubscribe();
    this.selectionsSubscription.unsubscribe();
  }

  // Helper methods
  loadAvailableCompanies(): void {
    this.gridState$.take(1).subscribe(gridState => {
      this.store.dispatch(new fromAvailableCompaniesActions.LoadingAvailableCompanies(
        {
          exchangeId: this.exchangeId,
          listState: gridState
        }
      ));
    });
  }
}
