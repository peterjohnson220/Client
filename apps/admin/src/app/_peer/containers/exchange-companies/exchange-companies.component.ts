import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';

import { ExchangeCompany } from 'libs/models';

import * as fromExchangeCompaniesActions from '../../actions/exchange-companies.actions';
import * as fromPeerAdminReducer from '../../reducers';

@Component({
  selector: 'pf-exchange-companies',
  templateUrl: './exchange-companies.component.html',
  styleUrls: ['./exchange-companies.component.scss']
})
export class ExchangeCompaniesComponent implements OnInit {
  exchangeCompaniesLoading$: Observable<boolean>;
  exchangeCompaniesLoadingError$: Observable<boolean>;
  exchangeCompanies$: Observable<ExchangeCompany[]>;
  exchangeId: number;

  constructor(
    private store: Store<fromPeerAdminReducer.State>,
    private route: ActivatedRoute
  ) {
    this.exchangeCompaniesLoading$ = this.store.select(fromPeerAdminReducer.getExchangeCompaniesLoading);
    this.exchangeCompaniesLoadingError$ = this.store.select(fromPeerAdminReducer.getExchangeCompaniesLoadingError);
    this.exchangeCompanies$ = this.store.select(fromPeerAdminReducer.getExchangeCompanies);

    this.exchangeId = this.route.snapshot.params.id;
  }

  // Events
  handleExchangeCompaniesGridReload() {
    this.store.dispatch(new fromExchangeCompaniesActions.LoadingExchangeCompanies(this.exchangeId));
  }

  // Lifecycle
  ngOnInit() {
    this.store.dispatch(new fromExchangeCompaniesActions.LoadingExchangeCompanies(this.exchangeId));
  }
}
