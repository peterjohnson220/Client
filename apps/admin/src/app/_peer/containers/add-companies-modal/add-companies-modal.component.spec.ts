import { forwardRef, NO_ERRORS_SCHEMA } from '@angular/core';
import { NG_VALUE_ACCESSOR, ReactiveFormsModule } from '@angular/forms';
import { ActivatedRoute } from '@angular/router';
import { TestBed, ComponentFixture, async, fakeAsync, tick } from '@angular/core/testing';

import { combineReducers, Store, StoreModule } from '@ngrx/store';
import { of } from 'rxjs/observable/of';

import { PfCommonModule } from 'libs/common';
import { PfValidatableDirective } from 'libs/forms/directives';
import * as fromRootState from 'libs/state/state';
import * as fromPeerAdminReducer from '../../reducers';
import * as fromAvailableCompaniesActions from '../../actions/available-companies.actions';
import { AddCompaniesModalComponent } from './add-companies-modal.component';
import { generateMockAddExchangeCompaniesRequest } from '../../../../../../../libs/models/peer';
import * as fromExchangeCompaniesActions from '../../actions/exchange-companies.actions';
import { GridFilterService } from '../../../../../../../libs/common/core/services';
import { PfInputDebounceComponent } from '../../../../../../../libs/forms/components';

describe('Add Companies Modal', () => {
  let fixture: ComponentFixture<AddCompaniesModalComponent>;
  let instance: AddCompaniesModalComponent;
  // let pfInputDebounce: PfInputDebounceComponent;
  let store: Store<fromRootState.State>;
  let activatedRoute: ActivatedRoute;
  let routeIdParam: number;

  // Configure Testing Module for before each test
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          peerAdmin: combineReducers(fromPeerAdminReducer.reducers)
        }),
        ReactiveFormsModule,
        PfCommonModule
      ],
      declarations: [
        AddCompaniesModalComponent,
        PfValidatableDirective
      ],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: { snapshot: { params: { id : 1 } } },
        },
        {
          provide: PfInputDebounceComponent,
          // useValue: jest.mock(PfInputDebounceComponent)
          useValue: {setSilently(test: string) {}}
        }
      ],
      // Shallow Testing
      schemas: [ NO_ERRORS_SCHEMA ]
    });

    store = TestBed.get(Store);
    activatedRoute = TestBed.get(ActivatedRoute);
    // pfInputDebounce = TestBed.createComponent(PfInputDebounceComponent).componentInstance;
    routeIdParam = activatedRoute.snapshot.params.id;

    spyOn(store, 'dispatch');

    fixture = TestBed.createComponent(AddCompaniesModalComponent);
    instance = fixture.componentInstance;
    instance.debouncedSearchTerm = TestBed.get(PfInputDebounceComponent);
  });

  it('should show a modal with a search bar and a companies grid when addCompaniesModalOpen$ is true', () => {
    instance.addCompaniesModalOpen$ = of(true);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a LoadingAvailableCompanies action when the modal is opened', () => {
    const action = new fromAvailableCompaniesActions.LoadingAvailableCompanies({
      exchangeId: instance.exchangeId,
      listState: JSON.stringify(instance.gridState)
    });
    instance.addCompaniesModalOpen$ = of(true);
    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should call selectionsControl.setErrors when addingCompaniesError$ is true', () => {
    const error = {'error': 'There was an error adding the selected companies.'};
    spyOn(instance.selectionsControl, 'setErrors');
    instance.addingCompaniesError$ = of(true);

    fixture.detectChanges();

    expect(instance.selectionsControl.setErrors).toHaveBeenCalledWith(error);
  });

  it('should dispatch a AddingExchangeCompanies action when the handleFormSubmit event is triggered', () => {
    const payload = generateMockAddExchangeCompaniesRequest();
    const action = new fromExchangeCompaniesActions.AddingExchangeCompanies(payload);

    instance.selections = [1, 2];

    fixture.detectChanges();

    instance.handleFormSubmit();

    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should have attemptedSubmit of true once handleFormSubmit is triggered', () => {
    instance.attemptedSubmit = false;
    fixture.detectChanges();

    instance.handleFormSubmit();

    fixture.detectChanges();

    expect(instance.attemptedSubmit).toBe(true);
  });

  it('should have attemptedSubmit of false once handleModalDismissed is triggered', () => {
    instance.attemptedSubmit = true;
    fixture.detectChanges();

    instance.handleModalDismissed();

    fixture.detectChanges();

    expect(instance.attemptedSubmit).toBe(false);
  });

  it('should dispatch a CloseAddExchangeCompaniesModal event when handleModalDismissed is triggered', () => {
    const action = new fromExchangeCompaniesActions.CloseAddExchangeCompaniesModal;
    fixture.detectChanges();

    instance.handleModalDismissed();

    fixture.detectChanges();

    expect(store.dispatch).toBeCalledWith(action);
  });

  it('should call debouncedSearchTerm.setSilently when handleModalDismissed is triggered', () => {
    // instance.debouncedSearchTerm = pfInputDebounce;
    spyOn(instance.debouncedSearchTerm, 'setSilently');
    fixture.detectChanges();

    instance.handleModalDismissed();

    fixture.detectChanges();

    expect(instance.debouncedSearchTerm).toBeCalledWith('');
  });

  it('should reset gridState and selections when handleModalDismissed is triggered', () => {
    instance.selections = [1, 2];
    instance.gridState = GridFilterService.getMockGridState();
    fixture.detectChanges();

    instance.handleModalDismissed();

    fixture.detectChanges();

    expect(instance.selections).toEqual([]);
    expect(instance.gridState).toEqual(GridFilterService.getMockEmptyGridState());
  });

  it('should update gridState and call loadAvailableCompanies when updateSearchFilter is triggered', () => {
    const expectedGridState = GridFilterService.getMockGridState('CompanyName', 0);
    instance.gridState = GridFilterService.getMockEmptyGridState();
    instance.gridState.skip = 10;
    const action = new fromAvailableCompaniesActions.LoadingAvailableCompanies({
      exchangeId: instance.exchangeId,
      listState: JSON.stringify(instance.gridState)
    });
    fixture.detectChanges();

    instance.updateSearchFilter('test');

    fixture.detectChanges();

    expect(instance.gridState).toEqual(expectedGridState);
    expect(store.dispatch).toBeCalledWith(action);
  });

/*  it('should show error message when addingCompaniesError$ is true and a submit has been attempted', () => {
    spyOn(instance, 'addingCompaniesError$').and.returnValue(of(true));
    instance.selectionsControl.setValue([1]);
    instance.selectionsControl.markAsDirty();
    instance.selectionsControl.markAsTouched();

    // instance.addingCompaniesError$ = of(false);
    instance.addCompaniesModalOpen$ = of(true);
    instance.attemptedSubmit = true;

    fixture.detectChanges();
    instance.addCompaniesModalOpen$ = of(true);

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
    // instance.addCompaniesModalOpen$ = of(true);
    // instance.addingCompaniesError$ = of(true);

    // fixture.detectChanges();




    // expect(fixture).toMatchSnapshot();
  });*/

/*  it('should not show the form text and should show a required error message once a submit has been attempted', () => {
    fixture.detectChanges();

    // trigger handleFormSubmit
    instance.handleFormSubmit();

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch an CloseCreateExchangeModal action when handleModalDismissed is called', () => {
    const action = new fromExchangeListActions.CloseCreateExchangeModal();

    instance.handleModalDismissed();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch an UpsertingExchange action with payload when handleFormSubmit is called', () => {
    const action = new fromExchangeListActions.UpsertingExchange(generateMockUpsertExchangeRequest());
    instance.name.setValue('test');
    instance.handleFormSubmit();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });*/
});
