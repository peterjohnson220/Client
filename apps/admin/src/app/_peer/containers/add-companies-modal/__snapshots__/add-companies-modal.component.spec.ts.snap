// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Companies Modal should show a modal with a search bar and a companies grid when addCompaniesModalOpen$ is true 1`] = `
<pf-add-companies-modal
  addCompaniesErrorSubscription={[Function Subscriber]}
  addCompaniesForm={[Function FormGroup]}
  addCompaniesModalOpen$={[Function ScalarObservable]}
  addCompaniesModalOpenSubscription={[Function Subscriber]}
  addingCompanies$={[Function Store]}
  addingCompaniesError$={[Function Store]}
  attemptedSubmit="false"
  availableCompaniesLoading$={[Function Store]}
  availableCompaniesLoadingError$={[Function Store]}
  debouncedSearchTerm="undefined"
  exchangeId={[Function Number]}
  fb={[Function FormBuilder]}
  gridState={[Function Object]}
  route={[Function Object]}
  selections={[Function Array]}
  store={[Function Store]}
  view$={[Function Store]}
>
  <pf-modal-form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    title="Add Companies"
  >
    
  
    
    
    
  
      
    <pf-async-container>
      
  
        
      <div
        class="form-group"
      >
        
          
        <pf-input-debounce
          id="search-box"
        >
          
          
        </pf-input-debounce>
        
        
      </div>
      
  
        
      <kendo-grid
        class="row-click-selects"
      >
        
  
          
        <kendo-grid-column
          field="CompanyName"
          title="Company Name"
        />
        
  
          
        <kendo-grid-column
          field="Industry"
          title="Industry"
        />
        
  
          
        <kendo-grid-column
          field="Revenue"
          title="Revenue"
        >
          
            
          
          
          
        </kendo-grid-column>
        
  
          
        <kendo-grid-column
          field="FTEs"
          title="FTEs"
        />
        
  
          
        <kendo-grid-column
          field="City"
          title="City"
        />
        
  
          
        <kendo-grid-column
          field="State"
          title="State"
        />
        
  
          
        <kendo-grid-column
          field="Zip"
          title="Zip"
        />
        
  
        
      </kendo-grid>
      
  
      
    </pf-async-container>
    
  
      
    <div
      class="form-group"
    >
      
        
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        formcontrolname="selections"
        name="hasSelections"
        ng-reflect-name="selections"
        type="hidden"
        value=""
      />
      
        
      
      
      
    </div>
    
  
    
    
  
  
  </pf-modal-form>
  
  
</pf-add-companies-modal>
`;

exports[`Add Companies Modal should show error message when addingCompaniesError$ is true and a submit has been attempted 1`] = `
<pf-add-companies-modal
  addCompaniesErrorSubscription={[Function Subscriber]}
  addCompaniesForm={[Function FormGroup]}
  addCompaniesModalOpen$={[Function Store]}
  addCompaniesModalOpenSubscription={[Function Subscriber]}
  addingCompanies$={[Function Store]}
  addingCompaniesError$={[Function ScalarObservable]}
  attemptedSubmit={[Function Boolean]}
  availableCompaniesLoading$={[Function Store]}
  availableCompaniesLoadingError$={[Function Store]}
  debouncedSearchTerm="undefined"
  exchangeId={[Function Number]}
  fb={[Function FormBuilder]}
  gridState={[Function Object]}
  route={[Function Object]}
  selections={[Function Array]}
  store={[Function Store]}
  view$={[Function Store]}
>
  <pf-modal-form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    title="Add Companies"
  >
    
  
    
    
    
  
      
    <pf-async-container>
      
  
        
      <div
        class="form-group"
      >
        
          
        <pf-input-debounce
          id="search-box"
        >
          
          
        </pf-input-debounce>
        
        
      </div>
      
  
        
      <kendo-grid
        class="row-click-selects"
      >
        
  
          
        <kendo-grid-column
          field="CompanyName"
          title="Company Name"
        />
        
  
          
        <kendo-grid-column
          field="Industry"
          title="Industry"
        />
        
  
          
        <kendo-grid-column
          field="Revenue"
          title="Revenue"
        >
          
            
          
          
          
        </kendo-grid-column>
        
  
          
        <kendo-grid-column
          field="FTEs"
          title="FTEs"
        />
        
  
          
        <kendo-grid-column
          field="City"
          title="City"
        />
        
  
          
        <kendo-grid-column
          field="State"
          title="State"
        />
        
  
          
        <kendo-grid-column
          field="Zip"
          title="Zip"
        />
        
  
        
      </kendo-grid>
      
  
      
    </pf-async-container>
    
  
      
    <div
      class="form-group"
    >
      
        
      <input
        class="form-control ng-untouched ng-pristine ng-invalid"
        formcontrolname="selections"
        name="hasSelections"
        ng-reflect-name="selections"
        ng-reflect-pf-validatable="true"
        type="hidden"
        value=""
      />
      
        
      
      
      
          
      
      
        
      
      
    </div>
    
  
    
    
  
  
  </pf-modal-form>
  
  
</pf-add-companies-modal>
`;
